File: src/utils/data_structures/River_crossing_puzzles/PG.py
class PGState:
    def __init__(self, guards_left: int = 3, prisoners_left: int = 3, boat_side: str = "L"):
        self.guards_left = guards_left
        self.prisoners_left = prisoners_left
        self.boat_side = boat_side
        
    def __str__(self) -> str:
        return (f"{self.guards_left},{self.prisoners_left},{self.boat_side}")
        
def pg_result_of_move(state: PGState, num_guards: int, num_prisoners: int) -> Union[PGState, None]:
    if num_guards < 0 or num_prisoners < 0:
        return None
    if num_guards + num_prisoners == 0:
        return None
    if num_guards + num_prisoners > 2:
        return None
        
    G_L: int = state.guards_left
    G_R: int = (3 - state.guards_left)
    P_L: int = state.prisoners_left
    P_R: int = (3 - state.prisoners_left)
    if state.boat_side == "L":
        G_L -= num_guards
        G_R += num_guards
        P_L -= num_prisoners
        P_R += num_prisoners
        new_side: str = "R"
    else:
        G_L += num_guards
        G_R -= num_guards
        P_L += num_prisoners
        P_R -= num_prisoners
        new_side = "L"
        
    if G_L < 0 or P_L < 0 or G_R < 0 or P_R < 0:
        return None
        
    if G_L > 0 and G_L < P_L:
        return None
    if G_R > 0 and G_R < P_R:
        return None
    return PGState(G_L, P_L, new_side)

def pg_neighbors(state: PGState) -> list:
    neighbors: list = []
    for move in [(1,0), (2,0), (0,1), (0,2), (1,1)]:
        n: Union[PGState, None] = pg_result_of_move(state, move[0], move[1])
        if n is not None:
            neighbors.append(n)
    
    return neighbors
    
def create_prisoners_and_guards() -> Graph:
    indices: dict = {}
    next_node: queue.Queue = queue.Queue()
    g: Graph = Graph(0, undirected=True)
    
    initial_state: PGState = PGState(3, 3, "L")
    initial: Node = g.insert_node(label=initial_state)
    next_node.put(initial.index)
    indices[str(initial_state)] = initial.index
    
    while not next_node.empty():
        current_ind: int = next_node.get()
        current_node: Node = g.nodes[current_ind]
        current_state = current_node.label
        
        neighbors: list = pg_neighbors(current_state)
        for state in neighbors:
            state_str: str = str(state)
            if not state_str in indices:
                new_node: Node = g.insert_node(label=state)
                
                indices[state_str] = new_node.index
                next_node.put(new_node.index)
            new_ind: int = indices[str(state)]
            g.insert_edge(current_ind, new_ind, 1.0)
            
    return g
    
def pg_state_to_index_map(g: Graph) -> dict:
    state_to_index: dict = {}
    for node in g.nodes:
        state: str = str(node.label)
        state_to_index[state] = node.index
    return state_to_index
    
def solve_pg_bfs():
    g: Graph = create_prisoners_and_guards()
    
    state_to_index: dict = pg_state_to_index_map(g)
    state_index: int = state_to_index["3,3,L"]
    end_index: int = state_to_index["0,0,R"]
    last: int = breadth_first_search(g, start_index)
    
    current: int = end_index
    path_reversed: list = []
    while current != -1:
        path_reversed.append(current)
        current = last[current]
        
    if path_reversed[-1] != start_index:
        print("No solution")
        return
    
    for i, n in enumerate(reversed(path_reversed)):
        print(f"Step {i}: {g.nodes[n].label}")

File: src/utils/data_structures/River_crossing_puzzles/PG.ts
import { Graph2, breadth_first_search } from "../Graph2";

class PGState {
  guards_left: number;
  prisoners_left: number;
  boat_side: string;

  constructor(guards_left: number = 3, prisoners_left: number = 3, boat_side: string = "L") {
    this.guards_left = guards_left;
    this.prisoners_left = prisoners_left;
    this.boat_side = boat_side;
  }

  toString() {
    return `${this.guards_left},${this.prisoners_left},${this.boat_side}`;
  }
}

/**
 * TODO:
 * STUDY result_of_move
 * this can be used in front-end
 */
function pg_result_of_move(state: PGState, num_guards: number, num_prisoners: number) {
  if (num_guards < 0 || num_prisoners < 0) return null;
  if (num_guards + num_prisoners === 0) return null;
  if (num_guards + num_prisoners > 2) return null;

  /**
   * note:
   * computes the resulting number of prisoners and guards on both
   * left and right shores.
   */
  let G_L = state.guards_left;
  let G_R = 3 - state.guards_left;
  let P_L = state.prisoners_left;
  let P_R = 3 - state.prisoners_left;

  let new_side;
  if (state.boat_side === "L") {
    // num_guards: 1, 2, 0, 0, 1
    G_L -= num_guards;
    G_R += num_guards;
    // num_prisoners: 0, 0, 1, 2, 1
    P_L -= num_prisoners;
    P_R += num_prisoners;
    new_side = "R";
  } else {
    // num_guards: 1, 2, 0, 0, 1
    G_L += num_guards;
    G_R -= num_guards;
    // num_prisoners: 0, 0, 1, 2, 1
    P_L += num_prisoners;
    P_R -= num_prisoners;
    new_side = "L";
  }

  /**
   * 4 counts above are used to check whether the new state is valid.
   * checks that the move is not relating more people than are on the current
   * shower by confirming that none of the counts become negative.
   */
  if (G_L < 0 || P_L < 0 || G_R < 0 || P_R < 0) return null;

  /**
   * it also checks that the new state has a valid Balance of guards and prisoners.
   */
  if (G_L > 0 && G_L < P_L) return null;
  if (G_R > 0 && G_R < P_R) return null;

  return new PGState(G_L, P_L, new_side);
}

function pg_neighbors(state: PGState) {
  let neighbors = [];
  const moves = [[1, 0], [2, 0], [0, 1], [0, 2], [1, 1]];

  for (const move of moves) {
    const n = pg_result_of_move(state, move[0], move[1]);
    if (n !== null) {
      neighbors.push(n);
    }
  }
  return neighbors;
}

export function create_prisoners_and_guards() {
  /**
   * Note:
   * mimicking bfs
   */
  const visited = {} as any; // dictionary
  const queue = []; // queue
  const g = new Graph2(0, true);
  
  const initial_state = new PGState(3, 3, "L");
  const initial = g.insertNode(initial_state);

  queue.push(initial.index);
  visited[initial_state.toString()] = initial.index;
  
  while (queue.length > 0) {
    const current_ind = queue.shift();
    const current_node = g.nodes[current_ind];
    const current_state = current_node.label;
    
    /**
     * Note:
     * unlike prev bfs, we cannot rely on graph node's edgelist.
     * instead, we will use pg_neighbors() to check
     * got possible neighboring states
     */
    const neighbors = pg_neighbors(current_state);
    for (const state of neighbors) {
      const state_str = state.toString();
      /**
       * if we havent seen state yet, create a node in graph
       * and mark as seen
       */
      if (!(state_str in visited)) {
        const new_node = g.insertNode(state);

        visited[state_str] = new_node.index;
        queue.push(new_node.index);
      }

      /**
       * and create new edge between current node
       * and its neighbors
       */
      const new_ind = visited[state.toString()];
      g.insertEdge(current_ind, new_ind, 1.0);
    }
  }
  
  /**
   * return a graph that includes only states that are reachable
   * from initial state by valid moves.
   */
  return g;
}

/**
 * helper function for Puzzle Search
 * allows us to look up the start and goal indices without
 * tracing through the graph
 */
function pg_state_to_index_map(g: any) {
  const state_to_index = {} as any;

  for (const node of g.nodes) {
    state_to_index[node.label.toString()] = node.index;
  }
  return state_to_index;
}

export function solve_pg_bfs(check_end_index: string = "0,0,R") {
  const g = create_prisoners_and_guards();

  console.log(g)
  
  const state_to_index = pg_state_to_index_map(g);

  const state_index = state_to_index["3,3,L"];
  const end_index = state_to_index[check_end_index]; // prev "0,0,R"
  const last = breadth_first_search(g, state_index) as any;

  let current = end_index;
  const path_reversed = [];

  // reverses parent list from graph
  while (current !== -1) {
    path_reversed.push(current);
    current = last[current];
  }
  
  // checks if its a valid move
  if (path_reversed[path_reversed.length - 1] !== state_index) {
    console.log("No solution");
    return;
  }
  
  let res = '';

  for (let i = 0; i < path_reversed.length; i++) {
    const n = path_reversed[path_reversed.length - 1 - i];
    console.log(`Step ${i}: ${g.nodes[n].label}`);
    res = g.nodes[n].label;
  }

  return res;
}

